dbms/src/Storages/DeltaMerge/VersionChain/Common.cpp

```diff
@@ -0,0 +1,153 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Interpreters/Context.h>
+#include <Storages/DeltaMerge/DMContext.h>
+#include <Storages/DeltaMerge/File/DMFilePackFilter.h>
+#include <Storages/DeltaMerge/VersionChain/Common.h>
+
+namespace DB::DM
+{
+RSResults getRSResultsByRanges(const DMContext & dm_context, const DMFilePtr & dmfile, const RowKeyRanges & ranges)
+{
+    if (ranges.empty())
+        return RSResults(dmfile->getPacks(), RSResult::All);
+
+    return DMFilePackFilter::loadFrom(
+               dm_context,
+               dmfile,
+               /*set_cache_if_miss*/ true,
+               ranges,
+               EMPTY_RS_OPERATOR,
+               /*read_packs*/ {})
+        ->getHandleRes();
+}
+
+namespace
+{
+template <typename T>
+T getMaxValue(const MinMaxIndex & minmax_index, size_t i)
+{
+    if constexpr (std::is_same_v<T, Int64>) // For non-clustered index handle column
+        return minmax_index.getIntMinMax(i).second;
+    else if constexpr (std::is_same_v<T, String>) // For clustered index handle column
+        return minmax_index.getStringMinMax(i).second.toString();
+    else if constexpr (std::is_same_v<T, UInt64>) // For version column
+        return minmax_index.getUInt64MinMax(i).second;
+    else
+        static_assert(false, "Not support type");
+}
+
+template <typename T>
+T getMinValue(const MinMaxIndex & minmax_index, size_t i)
+{
+    if constexpr (std::is_same_v<T, Int64>) // For non-clustered index handle column
+        return minmax_index.getIntMinMax(i).first;
+    else if constexpr (std::is_same_v<T, String>) // For clustered index handle column
+        return minmax_index.getStringMinMax(i).first.toString();
+    else
+        static_assert(false, "Not support type");
+}
+
+std::pair<RSResults, UInt32> clipRSResults(const RSResults & rs_results)
+{
+    const auto start = std::find_if(rs_results.begin(), rs_results.end(), [](RSResult r) { return r.isUse(); });
+    if (start == rs_results.end())
+        return {};
+    const auto end = std::find_if(start, rs_results.end(), [](RSResult r) { return !r.isUse(); });
+    const auto start_pack_id = start - rs_results.begin();
+    return std::pair{RSResults(start, end), start_pack_id};
+}
+} // namespace
+
+std::pair<RSResults, UInt32> getClippedRSResultsByRange(
+    const DMContext & dm_context,
+    const DMFilePtr & dmfile,
+    const std::optional<RowKeyRange> & segment_range)
+{
+    if (!segment_range)
+        return std::make_pair(RSResults(dmfile->getPacks(), RSResult::All), 0);
+
+    const auto handle_res = getRSResultsByRanges(dm_context, dmfile, {*segment_range});
+    return clipRSResults(handle_res);
+}
```

Comment:
IMO, there is no need to define `getRSResultsByRanges` and `clipRSResults`, just copy those codes together, which is simple enough.

This brings some unnecessary checks like we have checked the `segment_range` in `getClippedRSResultsByRange`, but we need to check `ranges.empty()` again in `getRSResultsByRanges`.

Reply:
And the comments of `getClippedRSResultsByRange` are 

```
// Clip RSResults by removing the leading and trailing RSResult::None.
// Return the clipped RSResults and the pack_id of the first not RSResult::None.
// Because DMFile of ColumnFileBig only takes packs intersect with the `segment_range`.
```

But what we return is the first range of `isUse` packs. like:

```
{None, None, All, All, None, All, None} -> return {{All, All}, 2}
```